docsite/rst/developing_program_flow_modules.rst

```diff
@@ -0,0 +1,392 @@
+.. _flow_modules:
+
+=======
+Modules
+=======
+
+This is an in-depth dive into understanding Ansible's program flow to execute
+modules.  It will be of use to people working on the portions of the Core
+Ansible Engine that execute a module, it may be of interest to people writing
+Ansible Modules, and people simply wanting to use Ansible Modules will likely
+not find anything helpful here.
+
+.. _flow_types_of_modules:
+
+Types of Modules
+================
+
+Ansible supports several different types of modules in its code base.  Some of
+these are for backwards compatibility and others to enable flexibility.
+
+.. _flow_action_plugins:
+
+Action Plugins
+--------------
+
+Action Plugins look like modules to end users who are writing :term:`playbooks` but
+they're distinct entities for the purposes of this paper.  Action Plugins
+always execute on the controller. They sometimes are able to do all work there
+(for instance, the debug Action Plugin which prints some text for the user to
+see or the assert Action Plugin which can test whether several values in
+a playbook satisfy certain criteria.)
+
+More often, Action Plugins set up some values and environment on the controller
+and then invoke an actual module on the managed node that does something with
+these values.  An easy to understand version of this is the
+:ref:`template Action Plugin <template>`.  The
+:ref:`template Action Plugin <template>` takes values from the user to
+construct a file in a temporary location on the controller using variables
+from the playbook environment.  It then transfers the temporary file to
+a temporary file on the remote system.  After that, it invokes the
+:ref:`copy module <copy>` which operates on the remote system to move the file
+into its final location, set file permissions, and so on.
+
+.. _flow_new_style_modules:
+
+New-style Modules
+-----------------
+
+All of the modules that ship with Ansible fall into this category.
+
+New-style modules have the arguments to the module embedded inside of them in
+some manner.  Non-new-style modules have to copy a separate file over to the
+managed node which is less efficient as it requires two over-the-wire
+connections instead of only one.
```

Comment:
Non-new-style modules must copy a separate file over to the managed node, which is less efficient as it requires two over-the-wire connections instead of only one.

(changed "have to" to "must" and added a comma after "node")